diff --git a/src/sonic-swss/debian/rules b/src/sonic-swss/debian/rules
index 8412a6f1..b0b89a6f 100755
--- a/src/sonic-swss/debian/rules
+++ b/src/sonic-swss/debian/rules
@@ -39,13 +39,13 @@ endif
 override_dh_auto_configure:
 	dh_auto_configure -- $(configure_opts)
 	# Configure Rust build for countersyncd
-	cargo fetch
-	cargo update -p swss-common
+	#cargo fetch
+	#cargo update -p swss-common
 
 override_dh_auto_build:
-	dh_auto_build
+	make -j2
 	# Build and test countersyncd Rust project
-	cargo build --release
+	cargo build --release --locked
 
 override_dh_auto_install:
 	dh_auto_install --destdir=debian/swss
