diff --git a/data/debian/sonic-host-services-data.gnoi-shutdown.service b/data/debian/sonic-host-services-data.gnoi-shutdown.service
index cb50c5b..7934ea3 100644
--- a/data/debian/sonic-host-services-data.gnoi-shutdown.service
+++ b/data/debian/sonic-host-services-data.gnoi-shutdown.service
@@ -6,10 +6,10 @@ After=network-online.target database.service
 
 [Service]
 Type=simple
-ExecStartPre=/usr/bin/python3 /usr/local/bin/check_platform.py
+ExecCondition=/usr/bin/python3 /usr/local/bin/check_platform.py
 ExecStartPre=/bin/bash /usr/local/bin/wait-for-sonic-core.sh
 ExecStart=/usr/bin/python3 /usr/local/bin/gnoi_shutdown_daemon.py
-Restart=always
+Restart=on-failure
 RestartSec=5
 
 [Install]
