--- a/drivers/mmc/host/sdhci-xenon.c	2026-09-17 08:46:09.505399669 +0000
+++ b/drivers/mmc/host/sdhci-xenon.c	2026-09-17 10:27:43.704795001 +0000
@@ -255,6 +255,39 @@
 		return pltfm_host->clock;
 }
 
+static int xenon_set_dma_mask(struct sdhci_host *host)
+{
+	struct sdhci_pltfm_host *pltfm_host;
+	struct xenon_priv *priv;
+	struct mmc_host *mmc;
+	int ret;
+
+	pltfm_host = sdhci_priv(host);
+	priv = sdhci_pltfm_priv(pltfm_host);
+	mmc = host->mmc;
+
+	dev_dbg(mmc_dev(mmc), "xenon_set_dma_mask: hwver %x quirks2 %x\n",
+		priv->hw_version, host->quirks2);
+
+	if (host->quirks2 & SDHCI_QUIRK2_BROKEN_64_BIT_DMA)
+		host->flags &= ~SDHCI_USE_64_BIT_DMA;
+
+	if (priv->hw_version == XENON_AC5 &&
+	    host->quirks2 & SDHCI_QUIRK2_BROKEN_64_BIT_DMA) {
+		dev_info(mmc_dev(mmc),
+			 "xenon_set_dma_mask: setting mask to 34 bits\n");
+		ret = dma_set_mask_and_coherent(mmc_dev(mmc), DMA_BIT_MASK(34));
+	} else {
+		/* Reproduce the generic sdhci_add_host() mask selection for
+		 * every other Xenon controller.
+		 */
+		ret = dma_set_mask_and_coherent(mmc_dev(mmc),
+			DMA_BIT_MASK(host->flags & SDHCI_USE_64_BIT_DMA ? 64 : 32));
+	}
+
+	return ret;
+}
+
 static const struct sdhci_ops sdhci_xenon_ops = {
 	.voltage_switch		= xenon_voltage_switch,
 	.set_clock		= sdhci_set_clock,
@@ -263,6 +296,7 @@
 	.reset			= xenon_reset,
 	.set_uhs_signaling	= xenon_set_uhs_signaling,
 	.get_max_clock		= xenon_get_max_clock,
+	.set_dma_mask		= xenon_set_dma_mask,
 };
 
 static const struct sdhci_pltfm_data sdhci_xenon_pdata = {
@@ -577,6 +611,16 @@
 	pm_runtime_enable(&pdev->dev);
 	pm_suspend_ignore_children(&pdev->dev, 1);
 
+	/*
+	 * If we previously detected AC5 with over 2GB of memory,
+	 * then we disable ADMA and 64-bit DMA.
+	 * This means generic SDHCI driver has set the DMA mask to
+	 * 32-bit. Since DDR starts at 0x2_0000_0000, we must use
+	 * 34-bit DMA mask to access this DDR memory:
+	 */
+	if (priv->hw_version == XENON_AC5 &&
+	    host->quirks2 & SDHCI_QUIRK2_BROKEN_64_BIT_DMA)
+		dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(34));
 	err = sdhci_add_host(host);
 	if (err)
 		goto remove_sdhc;
