From 00edab837b25bb0d5379ec861710104c9be8d137 Mon Sep 17 00:00:00 2001
From: Chirag Shah <chirag@nvidia.com>
Date: Wed, 12 Nov 2025 16:13:11 -0800
Subject: [PATCH 02/29] zebra: EVPNv6 MH remote es parse ipaddr

Signed-off-by: Chirag Shah <chirag@nvidia.com>
---
 zebra/zebra_evpn_mh.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/zebra/zebra_evpn_mh.c b/zebra/zebra_evpn_mh.c
index 7d8e80cbc6..53b2653490 100644
--- a/zebra/zebra_evpn_mh.c
+++ b/zebra/zebra_evpn_mh.c
@@ -2575,9 +2575,7 @@ void zebra_evpn_proc_remote_es(ZAPI_HANDLER_ARGS)
 	s = msg;
 
 	STREAM_GET(&esi, s, sizeof(esi_t));
-	/* Temporary until BGP supports IPv6 VTEP */
-	SET_IPADDR_V4(&vtep_ip);
-	STREAM_GET(&vtep_ip.ipaddr_v4.s_addr, s, sizeof(vtep_ip.ipaddr_v4.s_addr));
+	STREAM_GET_IPADDR(s, &vtep_ip);
 
 	if (hdr->command == ZEBRA_REMOTE_ES_VTEP_ADD) {
 		uint32_t zapi_flags;
-- 
2.35.6

