From ca51d39440db46e766a81987f407ef794d697876 Mon Sep 17 00:00:00 2001
From: dgsudharsan <sudharsand@nvidia.com>
Date: Fri, 8 Nov 2024 22:14:31 +0000
Subject: [PATCH] Updating rules to set multipath to 512 and removing bgp vnc


diff --git a/debian/rules b/debian/rules
index aa13106fec..e280b6c45b 100755
--- a/debian/rules
+++ b/debian/rules
@@ -68,8 +68,7 @@ override_dh_auto_configure:
 		--disable-protobuf \
 		--disable-zeromq \
 		--enable-ospfapi \
-		--enable-bgp-vnc \
-		--enable-multipath=256 \
+		--enable-multipath=512 \
 		\
 		--enable-user=frr \
 		--enable-group=frr \
-- 
2.43.2

