From 2520a46942da5c65de1723af0df2885eb56923ea Mon Sep 17 00:00:00 2001
From: philo <philo@micasnetworks.com>
Date: Thu, 31 Oct 2024 03:45:04 -0500
Subject: [PATCH] zebra: Prevent protodown_rc from going Bzonkas

Signed-off-by: philo <philo@micasnetworks.com>
---
 zebra/interface.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/zebra/interface.c b/zebra/interface.c
index 85aaf0d46..fd6c68fc7 100644
--- a/zebra/interface.c
+++ b/zebra/interface.c
@@ -1187,6 +1187,12 @@ static bool if_ignore_set_protodown(const struct interface *ifp, bool new_down,
 
 	zif = ifp->info;
 
+	/*
+	 * FRR does not have enough data to make this request
+	 */
+	if (ifp->ifindex == IFINDEX_INTERNAL)
+		return true;
+
 	/* Current state as we know it */
 	old_down = !!(ZEBRA_IF_IS_PROTODOWN(zif));
 	old_set_down = !!CHECK_FLAG(zif->flags, ZIF_FLAG_SET_PROTODOWN);
-- 
2.25.1

